core/db/migrate/20190106184413_remove_code_from_spree_promotions.rb

```diff
@@ -0,0 +1,89 @@
+# frozen_string_literal: true
+
+class RemoveCodeFromSpreePromotions < ActiveRecord::Migration[5.1]
+  class Spree::Promotion < ActiveRecord::Base
+    has_many :codes, class_name: "Spree::PromotionCode"
+  end
+
+  def up
+    promotions_with_code = Spree::Promotion.where.not(code: nil)
+
+    if promotions_with_code.any?
+      # You have some promotions with "code" field present! This is not good
+      # since we are going to remove that column.
+      #
+      self.class.promotions_with_code_handler.new(self, promotions_with_code).call
```

Comment:
I think it would be better to use a simple callable instead of a class that responds to `#call` and pass options directly to `call`, also a constant would be more straightforward than a method:

```suggestion

      deprecated_code_handler = StopWithError
      # deprecated_code_handler = DoNothing
      # deprecated_code_handler = MoveToSpreePromotionCode
      deprecated_code_handler.call(self, promotions_with_code)
```

and then below:

```rb
# Please note that this will block the current migration and rollback all
# the previous ones run with the same "rails db:migrate" command.
StopWithError = ->(migration, promotions) {
  raise StandardError, "You are trying to drop 'code' column from "\
    "spree_promotions table but you have at least one record with that "\
    "column filled. Please take care of that or you could lose data. See:" \
    "\n" \
    "https://github.com/solidusio/solidus/pull/3028"\
    "\n"
}
```

This is much simpler and has the advantage of having the configuration near the error location, making it easier to find

Reply:
This is very similar to what I did first but to test all the possible proposed approaches I ended up creating a method. I know it's not ideal but I can't think to other solutions for this. Feel free to suggest how can I keep this code tested reducing its complexity.

Reply:
I like Elias idea of using simple callables, but Albertos approach is working well and is tested (not that easy with lambdas), so I think it's fine. The outcome is the same. 

What I do care about is the very thoughtful work that is done here and want to give an extra appreciation! Very nicely done, Alberto! 👏